File: gateway/src/pipeline/plugins/cors.ts
import type { ExecContext, PluginDeps } from '../types.js'
import { abort } from '../types.js'

export async function cors(ctx: ExecContext, deps: PluginDeps): Promise<void> {
  const origin = ctx.requestHeaders['origin'] ?? ctx.requestHeaders['Origin']
  if (!origin) return // 非 CORS 請求

  const allowedOrigins = parseCSV(deps.config['allowed_origins'] ?? '*')
  const allowedMethods = parseCSV(deps.config['allowed_methods'] ?? 'GET,POST,PUT,DELETE,OPTIONS')
  const allowedHeaders = parseCSV(deps.config['allowed_headers'] ?? 'Content-Type,Authorization')
  const exposedHeaders = deps.config['exposed_headers']
  const allowCreds     = deps.config['allow_credentials'] === 'true'
  const maxAge         = deps.config['max_age'] ?? '3600'

  if (!originMatches(origin, allowedOrigins)) {
    abort(ctx, 403, 'CORS: origin not allowed')
    return
  }

  ctx.responseHeaders['access-control-allow-origin'] = origin
  ctx.responseHeaders['vary'] = 'Origin'
  if (allowCreds) ctx.responseHeaders['access-control-allow-credentials'] = 'true'
  if (exposedHeaders) ctx.responseHeaders['access-control-expose-headers'] = exposedHeaders

  // Preflight
  if (ctx.method.toUpperCase() === 'OPTIONS') {
    ctx.responseHeaders['access-control-allow-methods'] = allowedMethods.join(', ')
    ctx.responseHeaders['access-control-allow-headers'] = allowedHeaders.join(', ')
    ctx.responseHeaders['access-control-max-age']       = maxAge
    abort(ctx, 204, '')
  }
}

function originMatches(origin: string, allowed: string[]): boolean {
  for (const a of allowed) {
    if (a === '*' || a.toLowerCase() === origin.toLowerCase()) return true
    if (a.startsWith('*.')) {
      const suffix = a.slice(1) // .example.com
      if (origin.toLowerCase().endsWith(suffix.toLowerCase())) return true
    }
  }
  return false
}

function parseCSV(s: string): string[] {
  return s.split(',').map((x) => x.trim()).filter(Boolean)
}

File: policy-engine/internal/plugins/cors.go
package plugins

import (
	"context"
	"net/http"
	"strconv"
	"strings"

	"github.com/xcloudapim/policy-engine/internal/domain"
)

// CORSPlugin 處理跨來源資源共享（CORS）
// 支援 Preflight 快取、Credentials、動態 Origin 驗證
type CORSPlugin struct{}

func NewCORSPlugin() *CORSPlugin { return &CORSPlugin{} }

func (p *CORSPlugin) Type() domain.PolicyType { return domain.PolicyTypeCORS }

func (p *CORSPlugin) Validate(config map[string]string) []string {
	if cfgGet(config, "allowed_origins") == "" {
		return []string{"allowed_origins is required"}
	}
	return nil
}

func (p *CORSPlugin) Execute(ctx context.Context, execCtx *domain.ExecContext, config map[string]string) error {
	origin := execCtx.GetHeader("Origin")
	if origin == "" {
		return nil // 非 CORS 請求，跳過
	}

	allowedOrigins := parseCSV(cfgGetDefault(config, "allowed_origins", "*"))
	allowedMethods := parseCSV(cfgGetDefault(config, "allowed_methods", "GET,POST,PUT,DELETE,OPTIONS"))
	allowedHeaders := parseCSV(cfgGetDefault(config, "allowed_headers", "Content-Type,Authorization"))
	exposedHeaders := cfgGet(config, "exposed_headers")
	allowCredentials := cfgGet(config, "allow_credentials") == "true"
	maxAge := cfgGetDefault(config, "max_age", "3600")

	// 驗證 Origin
	originAllowed := originMatches(origin, allowedOrigins)
	if !originAllowed {
		execCtx.Abort(403, "CORS: origin not allowed")
		return nil
	}

	execCtx.SetResponseHeader("Access-Control-Allow-Origin", origin)
	execCtx.SetResponseHeader("Vary", "Origin")

	if allowCredentials {
		execCtx.SetResponseHeader("Access-Control-Allow-Credentials", "true")
	}
	if exposedHeaders != "" {
		execCtx.SetResponseHeader("Access-Control-Expose-Headers", exposedHeaders)
	}

	// Preflight（OPTIONS）
	if execCtx.Method == http.MethodOptions {
		execCtx.SetResponseHeader("Access-Control-Allow-Methods", strings.Join(allowedMethods, ", "))
		execCtx.SetResponseHeader("Access-Control-Allow-Headers", strings.Join(allowedHeaders, ", "))
		if _, err := strconv.Atoi(maxAge); err == nil {
			execCtx.SetResponseHeader("Access-Control-Max-Age", maxAge)
		}
		execCtx.Abort(http.StatusNoContent, "")
		return nil
	}

	return nil
}

func originMatches(origin string, allowed []string) bool {
	for _, a := range allowed {
		if a == "*" || strings.EqualFold(a, origin) {
			return true
		}
		// wildcard subdomain: *.example.com
		if strings.HasPrefix(a, "*.") {
			suffix := a[1:] // .example.com
			if strings.HasSuffix(strings.ToLower(origin), strings.ToLower(suffix)) {
				return true
			}
		}
	}
	return false
}

func parseCSV(s string) []string {
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if t := strings.TrimSpace(p); t != "" {
			out = append(out, t)
		}
	}
	return out
}
